File: backend/app/schemas/review.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, Literal

from pydantic import BaseModel, Field, ConfigDict, field_validator


class ReviewUserLite(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    first_name: str
    last_name: str
    email: str
    role: str
    is_active: bool


class RequestReviewCreate(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    comment: Optional[str] = Field(default=None, max_length=1500)

    @field_validator("comment", mode="before")
    @classmethod
    def clean_comment(cls, v):
        if v is None:
            return None
        s = str(v).strip()
        return s or None


class RequestReviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    request_id: int
    reviewer_user_id: int
    reviewee_user_id: int
    reviewer_role: Literal["CUSTOMER", "WORKER"]
    rating: int
    comment: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    reviewer: Optional[ReviewUserLite] = None
    reviewee: Optional[ReviewUserLite] = None


class RequestReviewSummaryOut(BaseModel):
    request_id: int
    status: str

    customer_review_done: bool = False
    worker_review_done: bool = False

    can_review_as_customer: bool = False
    can_review_as_worker: bool = False

    my_review: Optional[RequestReviewOut] = None
    customer_review: Optional[RequestReviewOut] = None
    worker_review: Optional[RequestReviewOut] = None

File: src/app/core/models/review.ts
export interface ReviewUserLite {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  role: string;
  is_active: boolean;
}

export interface RequestReviewCreate {
  rating: number;
  comment?: string | null;
}

export interface RequestReview {
  id: number;
  request_id: number;
  reviewer_user_id: number;
  reviewee_user_id: number;
  reviewer_role: 'CUSTOMER' | 'WORKER';
  rating: number;
  comment?: string | null;
  created_at: string;
  updated_at: string;
  reviewer?: ReviewUserLite | null;
  reviewee?: ReviewUserLite | null;
}

export interface RequestReviewSummary {
  request_id: number;
  status: string;

  customer_review_done: boolean;
  worker_review_done: boolean;

  can_review_as_customer: boolean;
  can_review_as_worker: boolean;

  my_review?: RequestReview | null;
  customer_review?: RequestReview | null;
  worker_review?: RequestReview | null;
}
